crates/core/tedge_api/goals.md

```diff
@@ -0,0 +1,363 @@
+
+Messages conversations are also asynchronous, meaning that upon receiving a
+message ThinEdge might not reply. (This does not preclude the transport layer
+to assure that the message has been well received)
+
+This has two reasons:
+
+- It makes communication easier to implement
+- It reflects the network situation
+
+Using such an interface is still clunky without some additional help.
+
+As all messages have an ID, it could be possible to design a way of 'awaiting'
+a response. This is left as a future addition.
+
+## Plugins
+
+_Note: 'Plugins' is simply a working name for now. We are free to rename this
+to 'Extension' or any other name we think is good._
+
+Plugins are what makes ThinEdge work! They receive messages and execute their
+specific action if applicable.
+
+This part is split into two parts: PluginBuilders and Plugins themselves.
+
+### PluginBuilders
+
+A `PluginBuilder` is a Rust struct that implements `PluginBuilder`. They are the
+sole sources of plugin kinds in ThinEdge. (This is not a limitation for
+proprietary and non-rust plugins, read further to see how those are handled.)
+They are registered at startup and hardcoded into the ThinEdge binary.
+
+### Plugins
+
+A Plugin is a Rust struct that implements `Plugin`. It is clear that requiring
+all users who wish to extend ThinEdge to learn and use Rust is a _bad idea_. To
+make sure those users also get a first class experience special kinds of
+plugins are bundled per default: `stdio-external`, `http-external`. (The
+specific list will surely evolve over time.) These get instantiated like any
+other with the exception that it is expected that all messages are forwarded,
+e.g. to another binary or http endpoint. This way any user simply needs to know
+what kind of messages exists and how to send/receive them, giving them the same
+situation like a pure-Rust plugin would have.
+
+# Drawbacks
+
+- Tying the software specification to Rust code makes it more brittle to
+  accidental incompatibilities
+- People might get confused as to how Plugins can be written -> They see "Rust"
+  and think they need to know the language to make their own
+- Having everything in a single process group can make it harder to correctly
+  segment security rules for individual plugins
+
+# Rationale and alternatives
+
+Playing to Rust's strengths makes for better maintainable software. Using
+Traits to specify behaviour and "plain old Rust objects" also make the messages
+clear. As ThinEdge is a specialized piece of software, this should be embraced
+to deliver on an amazing experience.
+
+Alternatives could include:
+
+- Defining the objects externally, with for example CapNProto
+- Defining the objects through an API like Swagger or similar
+- Not doing this
+
+# Unresolved questions
+
+- How to extend the `MessageKind` type?
+    - The enum itself is `#[non_exhaustive]`, but extending it still requires a
+      whole developer story
+    - What is the process of adding new variants?
```

Comment:
I have a proposal to address this point.

I think that we need to leverage the facts that the `PluginBuilder`s are known at compile time and that the need for new `MessageKind`s will come from new implementations of `PluginBuilder`.

Hence, I propose to let the `PluginBuilder`s define their `MessageKind`s.

```rust
trait MessageKind: Display, Debug, Serialize, Deserialize, Clone, Eq, PartialEq

pub struct Message<MessageKind> {
    origin: Address,
    destination: Address,
    kind: Kind,
    id: uuid::Uuid,
}

#[async_trait]
pub trait PluginBuilder: Sync + Send + 'static {

   // The kind of messages handled by instances of this plugin
   type MessageKind : MessageKind

   /// ...
}
```

For that to work, one need to be able to connect the `PluginBuilder`s  in a type-safe manner. If a plugin `A` needs to send messages to a plugin `B` named "myB", then a typed connection has to be established during initialization. What I propose is to have the `ABuilder`  asking the `BBuilder` for the instance of `B` named "myB". So the `ABuilder` can build an `A` with a property `A.myB` of type `B`.

The plugins only needs to send messages to their dependencies. Hence, an approach could be to use `MessageChannel` rather then reference to plugins. 
```rust
#[async_trait]
pub trait PluginBuilder: Sync + Send + 'static {
    // ...
    
    // Return a message channel to the instance with the given name.
    pub fn get_instance_channel(&self, name: &str) -> Result<MessageChannel<Self::MessageKind>, PluginError>
}
```

This is just a sketch, but I'm convinced that this opens a path to an extensible and typed-safe message set.

Reply:
I welcome any push towards more compile-time assurances!

Would this replace the runtime configuration? If not, how to reconcile the two sides (since type information is gone once compiled).

Since the core has already been written in a small form in this PR, I would love to see how your approach could look like!